File: src/commands/moderation/kick.ts
import type { GuildMember } from "discord.js";
import { Kick } from "../../lib/util/moderation/valeriyya.moderation.kick";
import { ValeriyyaEmbed } from "../../lib/util/valeriyya.embed";
import { defineCommand, type ICommandInteraction, OptionTypes } from "../../lib/util/valeriyya.types";

export default defineCommand({
    data: {
        name: "kick",
        description: "Kicks a member from the guild.",
        options: [
            {
                name: "member",
                description: "The member to kick.",
                type: OptionTypes.USER,
                required: true
            },
            {
                name: "reason",
                description: "The reason for this ban.",
                type: OptionTypes.STRING
            }
        ]
    },
    chat: async (int: ICommandInteraction) => {
        const staff = int.member as GuildMember;
        const target = int.options.getMember("member") as GuildMember;
        const reason = int.options.getString("reason") ?? "";

        if (!target.kickable) return {
            embeds: [
                new ValeriyyaEmbed(undefined, "error")
                .setDescription("I can't kick this person due to me being unable to manage him. (They have a higher role than me or they are the owner).")
                .setAuthor({ name: `${int.user.tag} (${int.user.id})`, url: int.user.displayAvatarURL({ dynamic: true }) })
            ],
            ephemeral: true
        }

        const date = Date.now();
        const action = new Kick({
            int,
            staff,
            target,
            date,
            reason
        });

        await action.all();

        const embed = new ValeriyyaEmbed()
            .setAuthor({ name: `${int.user.tag} (${int.user.id})`, iconURL: int.user.displayAvatarURL({dynamic: true}) })
            .setThumbnail(int.guild?.iconURL({dynamic: true}) ?? '')
            .setDescription(`${target} has been kicked from ${int.guild?.name}`);

        return {
            embeds: [embed],
            ephemeral: true
        }
    }
})

File: src/commands/moderation/kick.rs
use poise::serenity_prelude::{ChannelId, Timestamp, Member};

use crate::{
    structs::{ActionTypes, Case},
    utils::{member_managable,Valeriyya},
    Context, Error,
};

#[doc = "Kicks a member from the guild."]
#[poise::command(
    slash_command,
    category = "Moderation",
    default_member_permissions = "KICK_MEMBERS"
)]
pub async fn kick(
    ctx: Context<'_>,
    #[description = "The member to kick"] member: Member,
    #[description = "The reason for this kick."]
    #[rest]
    reason: Option<String>,
) -> Result<(), Error> {
    let database = &ctx.data().database();
    let guild_id = ctx.guild_id().unwrap();

    let mut guild_db = Valeriyya::get_database(database, guild_id.to_string()).await;
    let case_number = guild_db.cases_number + 1;
    let reason_default = reason.unwrap_or_else(|| format!("Use /reason {} <...reason> to set a reason for this case.", case_number));

    if !member_managable(ctx, &member).await {
        ctx.send(Valeriyya::reply("The member can't be managed so you can't kick them!").ephemeral(true)).await?;
        return Ok(());
    }
    member
        .kick_with_reason(ctx.discord(), &reason_default)
        .await?;
    let icon_url = ctx
        .guild()
        .unwrap()
        .icon_url()
        .unwrap_or_else(|| String::from(""));

    let message = if guild_db.channels.logs.as_ref().is_some() {
        let sent_msg = ChannelId(guild_db.channels.logs.as_ref().unwrap().parse::<std::num::NonZeroU64>().unwrap())
            .send_message(ctx.discord(), Valeriyya::msg_reply().add_embed(
                Valeriyya::embed()
                    .author(Valeriyya::reply_author(format!(
                        "{} ({})",
                        ctx.author().tag(),
                        ctx.author().id
                    )).icon_url(ctx.author().face()))
                    .thumbnail(&icon_url)
                    .description(format!(
                        "Member: `{}`\nAction: `{:?}`\nReason: `{}`",
                        member.user.tag(),
                        ActionTypes::Kick,
                        reason_default
                    ))
                    .footer(Valeriyya::reply_footer(format!("Case {}", case_number)))
            )).await.expect("Guild log channel doesn't exist");

        Some(sent_msg.id.to_string())
    } else {
        None
    };

    guild_db = guild_db.add_cases(Case {
        id: case_number,
        action: ActionTypes::Kick,
        guild_id: guild_id.to_string(),
        staff_id: ctx.author().id.to_string(),
        target_id: member.user.id.to_string(),
        date: Timestamp::unix_timestamp(&Timestamp::now()),
        reason: reason_default.to_string(),
        message,
        expiration: None,
        reference: None,
    });

    ctx.say(format!("{} has been kicked by {}!", member, ctx.author())).await?;

    guild_db.execute(database).await;
    Ok(())
}
